detoxify/detoxify.py

```diff
@@ -13,21 +13,25 @@
 
 
 def get_model_and_tokenizer(
-    model_type, model_name, tokenizer_name, num_classes, state_dict
+    model_type, model_name, tokenizer_name, num_classes, state_dict, pretrained_model_path=None
 ):
     model_class = getattr(transformers, model_name)
     model = model_class.from_pretrained(
-        pretrained_model_name_or_path=None,
-        config=model_type,
+        pretrained_model_name_or_path=pretrained_model_path,
```

Comment:
This doesn't need to be changed, `pretrained_model_name_or_path ` should remain as `None` in order to not download the pretrained transformer weights that are only going to be overridden by the detoxify checkpoints.

Reply:
hmm, I have tested it and I gad to overwrite it because it first created a model then loads the weights from the checkpoint...
happy to edit it but what you suggested did not work for me :rabbit:

Reply:
Note that if you want to test it you need to turn off your internet and remove HF cache :]

Reply:
I checked and only tokenizers appear in `~/.cache/huggingface` 

```
/Users/jamesthewlis/.cache/huggingface//transformers/534479488c54aeaf9c3406f647aa2ec13648c06771ffe269edabebd4c412da1d.7f2721073f19841be16f41b0a70b600ca6b880c8f3df6f3535cbc704371bdfa4: JSON data
/Users/jamesthewlis/.cache/huggingface//transformers/534479488c54aeaf9c3406f647aa2ec13648c06771ffe269edabebd4c412da1d.7f2721073f19841be16f41b0a70b600ca6b880c8f3df6f3535cbc704371bdfa4.json: JSON data
/Users/jamesthewlis/.cache/huggingface//transformers/c1d7f0a763fb63861cc08553866f1fc3e5a6f4f07621be277452d26d71303b7e.20430bd8e10ef77a7d2977accefe796051e01bc2fc4aa146bc862997a1a15e79.json: JSON data
/Users/jamesthewlis/.cache/huggingface//transformers/3c61d016573b14f7f008c02c4e51a366c67ab274726fe2910691e2a761acf43e.37395cee442ab11005bcd270f3c34464dc1704b715b5d7d52b1a461abe3b9e4e.json: JSON data
/Users/jamesthewlis/.cache/huggingface//transformers/45c3f7a79a80e1cf0a489e5c62b43f173c15db47864303a55d623bb3c96f72a5.d789d64ebfe299b0e416afc4a169632f903f693095b4629a7ea271d5a0cf2c99.json: JSON data
/Users/jamesthewlis/.cache/huggingface//transformers/c1d7f0a763fb63861cc08553866f1fc3e5a6f4f07621be277452d26d71303b7e.20430bd8e10ef77a7d2977accefe796051e01bc2fc4aa146bc862997a1a15e79: JSON data
/Users/jamesthewlis/.cache/huggingface//transformers/3c61d016573b14f7f008c02c4e51a366c67ab274726fe2910691e2a761acf43e.37395cee442ab11005bcd270f3c34464dc1704b715b5d7d52b1a461abe3b9e4e: JSON data
/Users/jamesthewlis/.cache/huggingface//transformers/45c3f7a79a80e1cf0a489e5c62b43f173c15db47864303a55d623bb3c96f72a5.d789d64ebfe299b0e416afc4a169632f903f693095b4629a7ea271d5a0cf2c99: Unicode text, UTF-8 text
/Users/jamesthewlis/.cache/huggingface//transformers/45c3f7a79a80e1cf0a489e5c62b43f173c15db47864303a55d623bb3c96f72a5.d789d64ebfe299b0e416afc4a169632f903f693095b4629a7ea271d5a0cf2c99.lock: empty
/Users/jamesthewlis/.cache/huggingface//transformers/3c61d016573b14f7f008c02c4e51a366c67ab274726fe2910691e2a761acf43e.37395cee442ab11005bcd270f3c34464dc1704b715b5d7d52b1a461abe3b9e4e.lock: empty
/Users/jamesthewlis/.cache/huggingface//transformers/c1d7f0a763fb63861cc08553866f1fc3e5a6f4f07621be277452d26d71303b7e.20430bd8e10ef77a7d2977accefe796051e01bc2fc4aa146bc862997a1a15e79.lock: empty
/Users/jamesthewlis/.cache/huggingface//transformers/534479488c54aeaf9c3406f647aa2ec13648c06771ffe269edabebd4c412da1d.7f2721073f19841be16f41b0a70b600ca6b880c8f3df6f3535cbc704371bdfa4.lock: empty
```

This is with the original model, what model are you using? Do you see weight files under `~/.cache/huggingface`?

Reply:
as shown in the kernel, I am loading `roberta-base` which shall be the model used for your unbias checkpoint, correct?

Reply:
So these are the files downloaded with empty cache for the unbiased model:

```
$ rm -rf ~/.cache/huggingface                                                                                                                                                                                                  
$ python -c 'import detoxify; d=detoxify.Detoxify("unbiased")'                                                                                                                                                                 
Downloading: "https://github.com/unitaryai/detoxify/releases/download/v0.3-alpha/toxic_debiased-c7548aa0.ckpt" to /Users/jamesthewlis/.cache/torch/hub/checkpoints/toxic_debiased-c7548aa0.ckpt
100%|██████████████████████| 476M/476M [00:21<00:00, 23.5MB/s]
Downloading: 100%|██████████████████████| 481/481 [00:00<00:00, 194kB/s]
Downloading: 100%|██████████████████████| 899k/899k [00:00<00:00, 1.62MB/s]
Downloading: 100%|██████████████████████| 456k/456k [00:00<00:00, 986kB/s]
Downloading: 100%|██████████████████████| 1.36M/1.36M [00:00<00:00, 2.33MB/s]

$ find ~/.cache/huggingface -type f | xargs file       
                                                                                                                                                                
/Users/jamesthewlis/.cache/huggingface/transformers/d3ccdbfeb9aaa747ef20432d4976c32ee3fa69663b379deb253ccfce2bb1fdc5.d67d6b367eb24ab43b08ad55e014cf254076934f71d832bbab9ad35644a375ab.lock: empty
/Users/jamesthewlis/.cache/huggingface/transformers/733bade19e5f0ce98e6531021dd5180994bb2f7b8bd7e80c7968805834ba351e.35205c6cfc956461d8515139f0f8dd5d207a2f336c0c3a83b4bc8dca3518e37b.lock: empty
/Users/jamesthewlis/.cache/huggingface/transformers/d53fc0fa09b8342651efd4073d75e19617b3e51287c2a535becda5808a8db287.fc9576039592f026ad76a1c231b89aee8668488c671dfbe6616bab2ed298d730:      Unicode text, UTF-8 text, with very long lines (61874), with no line terminators
/Users/jamesthewlis/.cache/huggingface/transformers/cafdecc90fcab17011e12ac813dd574b4b3fea39da6dd817813efa010262ff3f.5d12962c5ee615a4c803841266e9c3be9a691a924f72d395d3a6c6c81157788b.json: JSON data
/Users/jamesthewlis/.cache/huggingface/transformers/d53fc0fa09b8342651efd4073d75e19617b3e51287c2a535becda5808a8db287.fc9576039592f026ad76a1c231b89aee8668488c671dfbe6616bab2ed298d730.json: JSON data
/Users/jamesthewlis/.cache/huggingface/transformers/cafdecc90fcab17011e12ac813dd574b4b3fea39da6dd817813efa010262ff3f.5d12962c5ee615a4c803841266e9c3be9a691a924f72d395d3a6c6c81157788b:      Unicode text, UTF-8 text
/Users/jamesthewlis/.cache/huggingface/transformers/d53fc0fa09b8342651efd4073d75e19617b3e51287c2a535becda5808a8db287.fc9576039592f026ad76a1c231b89aee8668488c671dfbe6616bab2ed298d730.lock: empty
/Users/jamesthewlis/.cache/huggingface/transformers/733bade19e5f0ce98e6531021dd5180994bb2f7b8bd7e80c7968805834ba351e.35205c6cfc956461d8515139f0f8dd5d207a2f336c0c3a83b4bc8dca3518e37b:      JSON data
/Users/jamesthewlis/.cache/huggingface/transformers/d3ccdbfeb9aaa747ef20432d4976c32ee3fa69663b379deb253ccfce2bb1fdc5.d67d6b367eb24ab43b08ad55e014cf254076934f71d832bbab9ad35644a375ab:      JSON data
/Users/jamesthewlis/.cache/huggingface/transformers/cafdecc90fcab17011e12ac813dd574b4b3fea39da6dd817813efa010262ff3f.5d12962c5ee615a4c803841266e9c3be9a691a924f72d395d3a6c6c81157788b.lock: empty
/Users/jamesthewlis/.cache/huggingface/transformers/733bade19e5f0ce98e6531021dd5180994bb2f7b8bd7e80c7968805834ba351e.35205c6cfc956461d8515139f0f8dd5d207a2f336c0c3a83b4bc8dca3518e37b.json: JSON data
/Users/jamesthewlis/.cache/huggingface/transformers/d3ccdbfeb9aaa747ef20432d4976c32ee3fa69663b379deb253ccfce2bb1fdc5.d67d6b367eb24ab43b08ad55e014cf254076934f71d832bbab9ad35644a375ab.json: JSON data
```

The first is the detoxify checkpoint. The `733bade19e5f0ce98e6531021dd` file turns out to be some config metadata (although not the weights) for roberta: https://huggingface.co/roberta-base/resolve/main/config.json and the other files are the tokenizers

So indeed it does need the network to download these small config files for the transformers even though the weights themselves are in our checkpoints. Unfortunately it looks like we haven't included that config information in our checkpoints so it does need to be supplied externally to work offline (unless we just hardcode them maybe?).

So for this PR the thing that is potentially confusing to people is having both `checkpoint` and `pretrained_model_path` arguments since they sound like they do the same thing. Not sure what the most elegant thing to do there is. Maybe something like `transformers_offline_base_model_path`?

Reply:
> So for this PR the thing that is potentially confusing to people is having both `checkpoint` and `pretrained_model_path` arguments since they sound like they do the same thing. Not sure what the most elegant thing to do there is. Maybe something like `transformers_offline_base_model_path`?

completely agree, it is confusing, but about the model construction, that eventually you main save the HF architecture and tokenizer inside the checkpoint as Flash does in [here](https://github.com/PyTorchLightning/lightning-flash/tree/fix/serialize_tokenizer) but that may mean more work so the quick fix is this one I guess :rabbit:

Reply:
So looks like after changing back `pretrained_model_name_or_path=None` and the new `pretrained_model_path` arg to something more clear - I'd suggest `transformers_offline_files_path`, things are good to go?

Reply:
not sure if I understand what you suggest to do, rename the new `pretrained_model_path` to `transformers_offline_files_path`?

Reply:
Yes, but also change `pretrained_model_name_or_path` back to `None`.

I just reinstalled everything and removed the HF cache and this works offline:
```python
def get_model_and_tokenizer(
    model_type, model_name, tokenizer_name, num_classes, state_dict, pretrained_model_path=None
):
    model_class = getattr(transformers, model_name)
    model = model_class.from_pretrained(
        pretrained_model_name_or_path=None,
        config=transformers_offline_files_path or model_type,
        num_labels=num_classes,
        state_dict=state_dict,
        local_files_only=transformers_offline_files_path is not None,
    )
    tokenizer = getattr(transformers, tokenizer_name).from_pretrained(
        transformers_offline_files_path or model_type,
        local_files_only= transformers_offline_files_path is not None,
    )

    return model, tokenizer
```

where `transformers_offline_files_path` is a folder containing the extra files the roberta-base model needs:

roberta-base
 ┣ config.json
 ┣ dict.txt
 ┣ merges.txt
 ┣ tokenizer.json
 ┗ vocab.json

Reply:
Hmm, but do you see what you have there... what is meant to be a path or name you drop, and in the second what is meant to be a name you force to be path...
* `pretrained_model_name_or_path` ignored
* `config` <- path or name

Reply:
is there any specific reason why you want to have `pretrained_model_name_or_path=None`?

Reply:
It seems redundant to have to download the HF weights in offline mode since they're going to be overwritten by the detoxify weights anyway.

According to the HF documentation:

**pretrained_model_name_or_path** (str or os.PathLike, optional) — Can be either:  
- None if you are both providing the configuration and state dictionary

**config** (Union[PretrainedConfig, str, os.PathLike], optional) — Can be either:
- a string or path valid as input to from_pretrained().

Reply:
> None if you are both providing the configuration and state dictionary

yes, but in your case, you are not doing so :]

Reply:
How so if we give the configuration (or path to it) and the state dict here?
```python
    model = model_class.from_pretrained(
        pretrained_model_name_or_path=None,
        config=transformers_offline_files_path or model_type,
        num_labels=num_classes,
        state_dict=state_dict,
        local_files_only=transformers_offline_files_path is not None,
    )
```

Reply:
so you say that "transformers offline files path" is NOT "pretrained model path"?

Reply:
yes, it's just the path to the other files needed i.e. config and tokenizer files, whereas `pretrained_model_name_or_path` would need the model weights too

Reply:
```suggestion
        pretrained_model_name_or_path=None,
        config=transformers_offline_files_path or model_type,
```